### PR TITLE
ECOM-4311: Added the functionality to only refund the order and skip revocation step

### PR DESCRIPTION
**Pull request by aamir-khan**
ECOM-4311:

Added the button to approve refund without revocation.

I will create the sandbox later and will share the link.

### CODE REVIEW

File: ecommerce/extensions/api/v2/views/refunds.py [hunk 1]
```diff
@@ -86,17 +86,19 @@ class RefundProcessView(generics.UpdateAPIView):
...
+        if action not in (APPROVE, DENY, REFUND_PAYMENT_ONLY):
             raise ParseError('The action [{}] is not valid.'.format(action))
 
         refund = self.get_object()
         result = False
 
-        if action == APPROVE:
-            result = refund.approve()
+        if action in (APPROVE, REFUND_PAYMENT_ONLY):
+            revoke_enrollment = True if action == APPROVE else False
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/refunds.py]: `revoke_fulfillment`. We sell more than enrollments. This change should be made everyplace you've typed "enrollment".

**Comment by clintonb** [comment]
This needs tests.

**Comment by aamir-khan** [comment]
@clintonb I added the tests. Please review now.
@adampalay, @awaisdar001, @mushtaqak  please review this PR.

File: ecommerce/extensions/refund/models.py [hunk 2]
```diff
@@ -193,9 +193,16 @@ def approve(self):
...
                 return False
 
-        if self.status in (REFUND.PAYMENT_REFUNDED, REFUND.REVOCATION_ERROR):
+        if self.status in (REFUND.PAYMENT_REFUNDED, REFUND.REVOCATION_ERROR) and revoke_fulfillment:
             self._revoke_lines()
 
+        if not revoke_fulfillment and self.status == REFUND.PAYMENT_REFUNDED:
+            logger.info("Skipping the revocation step for refund [%d].", self.id)
+            # Mark the status complete as it does not involve the revocation.
+            self.set_status(REFUND.COMPLETE)
```
Comment (awaisdar001) [inline ecommerce/extensions/refund/models.py:202]: Can we also have a test to check if refund status was successfully updated? Just to test the change in `approve` method.

File: ecommerce/extensions/api/v2/tests/views/test_refunds.py [hunk 3]
```diff
@@ -202,18 +203,43 @@ def test_invalid_action(self):
...
+                self.assertEqual(response.status_code, 200)
+                refund = Refund.objects.get(id=self.refund.id)
+                if decision == 'refund_payment_only':
+                    self.assertEqual(response.data['status'], refund.status)
+                    self.assertEqual(response.data['status'], "Complete")
+                    patched_log.info.assert_called_once_with(
+                        "Skipping the revocation step for refund [%d].", self.refund.id)
+                elif decision == "approve":
+                    self.assertEqual(response.data['status'], refund.status)
+                    self.assertEqual(response.data['status'], "Complete")
```
Comment (aamir-khan) [inline ecommerce/extensions/api/v2/tests/views/test_refunds.py]: @awaisdar001 The refund status is being checked here.

File: ecommerce/extensions/api/v2/tests/views/test_refunds.py [hunk 4]
```diff
@@ -202,18 +203,43 @@ def test_invalid_action(self):
...
+
+        mock_issue_credit.return_value = None
+
+        with mock.patch("ecommerce.extensions.fulfillment.api.revoke_fulfillment_for_refund",
+                        side_effect=mock_fulfillment):
+            with mock.patch("ecommerce.extensions.refund.models.logger") as patched_log:
+                response = self.put(decision)
+                self.assertEqual(response.status_code, 200)
+                refund = Refund.objects.get(id=self.refund.id)
+                if decision == 'refund_payment_only':
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_refunds.py]: These should be three separate test cases. In general, if you're using `if-else` to determine the outcome of your test case, split it up.

**Comment by aamir-khan** [comment]
@clintonb I split the tests. Please review now.

File: ecommerce/extensions/api/v2/tests/views/test_refunds.py [hunk 5]
```diff
@@ -202,18 +203,70 @@ def test_invalid_action(self):
         response = self.put('reject')
         self.assertEqual(response.status_code, 400)
 
-    @ddt.data('approve', 'deny')
-    def test_success(self, action):
+    @mock.patch('ecommerce.extensions.refund.models.Refund._issue_credit')
+    def test_success_approve(self, mock_issue_credit):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_refunds.py]: `test_success_with_payment_and_fulfillment`

Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_refunds.py]: Why did you modify the original tests? You should only be _adding_ tests. You can modify them to de-dupe code; but, I see no need for further changes.

File: ecommerce/extensions/api/v2/views/refunds.py [hunk 6]
```diff
@@ -86,17 +86,19 @@ class RefundProcessView(generics.UpdateAPIView):
     def update(self, request, *args, **kwargs):
         APPROVE = 'approve'
         DENY = 'deny'
+        REFUND_PAYMENT_ONLY = 'refund_payment_only'
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/refunds.py]: approve_payment_only

File: ecommerce/extensions/refund/models.py [hunk 7]
```diff
@@ -193,9 +193,16 @@ def approve(self):
                 self.set_status(REFUND.PAYMENT_REFUND_ERROR)
                 return False
 
-        if self.status in (REFUND.PAYMENT_REFUNDED, REFUND.REVOCATION_ERROR):
+        if self.status in (REFUND.PAYMENT_REFUNDED, REFUND.REVOCATION_ERROR) and revoke_fulfillment:
```
Comment (clintonb) [inline ecommerce/extensions/refund/models.py]: Reverse these operands. It's cheaper to check the value of `revoke_fulfillment` than it is to determine if `self.status` is in a list. You can shortcut the operation by performing the cheaper operation first.

File: ecommerce/templates/oscar/dashboard/partials/refund_buttons.html [hunk 8]
```diff
@@ -2,7 +2,10 @@
 
 {% if refund.can_approve %}
 <button type="button" class="btn btn-success" data-refund-id="{{ refund.id }}" data-decision="approve" data-action="process-refund">
-    {% trans "Approve" %}
+    {% trans "Approve Credit and Revoke" %}
+</button>
+<button type="button" class="btn btn-success" data-refund-id="{{ refund.id }}" data-decision="refund_payment_only" data-action="process-refund">
```
Comment (clintonb) [inline ecommerce/templates/oscar/dashboard/partials/refund_buttons.html]: Replace `btn-success` with `btn-info` to avoid confusing users.

**Comment by aamir-khan** [comment]
@adampalay @clintonb Please review this PR all comments have beeb addressed.

**Comment by clintonb** [comment]
Please update your branch and squash commits.

File: ecommerce/extensions/api/v2/views/refunds.py [hunk 9]
```diff
@@ -86,17 +86,19 @@ class RefundProcessView(generics.UpdateAPIView):
...
+        if action not in (APPROVE, DENY, APPROVE_PAYMENT_ONLY):
             raise ParseError('The action [{}] is not valid.'.format(action))
 
         refund = self.get_object()
         result = False
 
-        if action == APPROVE:
-            result = refund.approve()
+        if action in (APPROVE, APPROVE_PAYMENT_ONLY):
+            revoke_fulfillment = True if action == APPROVE else False
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/refunds.py]: `revoke_fulfillment = action == APPROVE`

**Comment by clintonb** [comment]
👍

**Comment by adampalay** [comment]
👍 (once you update and address clinton's comments)

**Pull request merged**